src/librustc_mir/transform/inline.rs

```diff
@@ -691,6 +691,14 @@ impl<'a, 'tcx> MutVisitor<'tcx> for Integrator<'a, 'tcx> {
         self.in_cleanup_block = false;
     }
 
+    fn visit_retag(&mut self, fn_entry: &mut bool, place: &mut Place<'tcx>, loc: Location) {
+        self.super_retag(fn_entry, place, loc);
+
+        // We have to patch all inlined retags to be aware that they are no longer
```

Comment:
Why is that a problem? Doesn't this mean inlining changes the rules of UB?

Reply:
It means inlining makes fewer things UB. Optimizations may do that.

Reply:
Without this change, inlining could *introduce* UB. That would surely be wrong.

It would introduce UB because when `fn_entry == true`, we push a "barrier" onto the borrow stack, which must not be popped before the function returns. But with inlining, the current function now is something much more long-lived than previously, meaning that "though shall not pop" is in effect for too long -> UB.